### ISSUE TITLE
spirv-val doesn't check for OpSelectionMerge before OpBranchConditional

### ISSUE THREAD

**Issue by baldurk** [KhronosGroup/SPIRV-Tools#1185]
As mentioned in the title, spirv-val doesn't print any errors if `OpSelectionMerge` is omitted before an `OpBranchConditional`, even though the shader declares the `Shader` capability which requires that for structured flow.

Example repro shader:

```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main" %gl_VertexIndex
               OpSource GLSL 430
               OpName %main "main"
               OpName %gl_VertexIndex "gl_VertexIndex"
               OpDecorate %gl_VertexIndex BuiltIn VertexIndex
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
        %int = OpTypeInt 32 1
%_ptr_Input_int = OpTypePointer Input %int
%gl_VertexIndex = OpVariable %_ptr_Input_int Input
      %int_5 = OpConstant %int 5
       %bool = OpTypeBool
       %main = OpFunction %void None %3
          %5 = OpLabel
          %9 = OpLoad %int %gl_VertexIndex
         %12 = OpSGreaterThan %bool %9 %int_5
; Error, this must not be omitted
; OpSelectionMerge %14 None
               OpBranchConditional %12 %13 %14
         %13 = OpLabel
               OpBranch %14
         %14 = OpLabel
               OpReturn
               OpFunctionEnd
```

**Comment by dneto0** [comment]
Yes, this check is not implemented.

It's not the case that an OpBranchConditional *always* be preceded by an OpSelectionMerge or OpLoopMerge.  A conditional branch implementing a break or continue is such a case.

**Comment by jeremy-lunarg** [comment]
FWIW, I'm finishing up a patch which validates that OpLoopMerge is followed by an OpBranch or OpBranchConditional and OpSelectionMerge is followed by an OpBranchConditional or OpSwitch. Should have it ready tomorrow.

**Comment by JasperNV** [comment]
Seems like this was accomplished in cd68f2b17642a76ed8f76b233ffce76f777be55a ? I assume this can be closed.

**Comment by afd** [comment]
This does not appear to be resolved, e.g. this shader validates:

```
                 OpCapability Shader
            %1 = OpExtInstImport "GLSL.std.450"
                 OpMemoryModel Logical GLSL450
                 OpEntryPoint Fragment %4 "main"
                 OpExecutionMode %4 OriginUpperLeft
                 OpSource ESSL 310
            %2 = OpTypeVoid
            %3 = OpTypeFunction %2
            %6 = OpTypeInt 32 1
            %7 = OpTypePointer Function %6
            %9 = OpConstant %6 3
           %11 = OpConstant %6 4
           %12 = OpTypeBool
           %19 = OpConstantTrue %12
           %16 = OpConstant %6 1
           %18 = OpConstant %6 2
            %4 = OpFunction %2 None %3
            %5 = OpLabel
                 OpBranch %14
           %14 = OpLabel
                 OpBranchConditional %19 %20 %21
           %20 = OpLabel
                 OpBranch %21
           %21 = OpLabel
                 OpBranch %15
           %15 = OpLabel
                 OpReturn
                 OpFunctionEnd
```

**Comment by alan-baker** [comment]
This should be addressed by #2962
